hydra-test-utils/src/Test/Hydra/Prelude.hs

```diff
@@ -137,3 +155,113 @@ newtype ReasonablySized a = ReasonablySized a
 
 instance Arbitrary a => Arbitrary (ReasonablySized a) where
   arbitrary = ReasonablySized <$> reasonablySized arbitrary
+
+-- Like 'coverTable', but construct the weight requirements generically from the
+-- label type-representation.
+--
+--     data MyCoverLabel = Case1 | Case2 deriving (Generic, Data, Show)
+--
+--     forAll arbitrary $ \(lbl :: MyCoverLabel, arg) ->
+--         myProp arg
+--         & genericCoverTable @UniformCoverage [lbl]
+--         & checkCoverage
+--
+--     ===
+--
+--     forAll arbitrary $ (lbl :: MyCoverLabel, arg) ->
+--        myProp arg
+--        & coverTable "MyCoverTable" [(Case1, 50), (Case2, 50)]
+--        & tabulate "MyCoverTable" [show lbl]
+--        & checkCoverage
+--
+genericCoverTable ::
+  forall distr a prop.
+  ( GCoverTableRequirements distr (Rep a)
+  , GLabelName (Rep a)
+  , Generic a
+  , Show a
+  , Testable prop
+  ) =>
+  [a] ->
+  prop ->
+  Property
+genericCoverTable xs =
+  coverTable tableName requirements . tabulate tableName (gLabelName . from <$> xs)
+ where
+  (tableName, requirements) = gCoverTableRequirements @distr (Proxy @(Rep a))
```

Comment:
I still don't get why we would want to take `[a]` as we seemingly always only pass a single `[mutation]`? The documentation of [`tabulate`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:tabulate) then states it is preferred over `label` if there is a varying amount of values to label.. but we have only one, so a single `label` should fit our purpose?

Furthermore, the `tableName` is always `""` in the current implementation of `gCoverTableRequirements` and parameterizing the distribution is YAGNI in my opinion.

If we could reduce the "noise" of generic type classes and instances down to a single `constructorNames :: Generic a => Proxy a -> [String]` label producing function, that would be preferred to me -> keep it simple. WDYT?

Reply:
> why we would want to take [a] as we seemingly always only pass a single [mutation]

For the moment indeed. Maybe YAGNI, but we had initially in mind to run _several mutations_ at once and combine them arbitrarily. Also, since this function is quite generic in the end (may be used for other test), I preferred keeping the API flexible instead of being needlessly restrictive. It works well for list, and it's also straightforward for the caller to wrap a single element in a singleton list.

> the tableName is always "" in the current implementation of gCoverTableRequirements

No! The tableName is set to the name of the overall wrapping type (here, `CloseMutation`). But the generic implementation is maybe a big hard to read. It's really a top-down structure with the `D1 _ _` constructor at the top; which is thus where the tableName is actually set. At lower levels, we collect other information but use a default, empty `""` table name (which is filled by the `D1 _ _` level).  

> and parameterizing the distribution is YAGNI in my opinion.

Arguably. It's more for the readability and for the sake o writing something a bit generic. It's not like it costs much effort to be frank :smile: 

> If we could reduce the "noise" of generic type classes and instances down to a single constructorNames :: Generic a => Proxy a -> [String]

Then I am curious as for how you'd write that function implementation without type classes :blush: ...